### PR TITLE
Add guided decoding for OpenAI API server

### PR DESCRIPTION
**Pull request by felixzhu555**
Support guided decoding (JSON, regex, choice) using [outlines](https://github.com/outlines-dev/outlines/tree/main) for the completion and chat completion OpenAI endpoints.
This is a continuation of @br3no's work in #2815.

relevant: #288

### CODE REVIEW

**Comment by jalotra** [comment]
@felixzhu555 can you please add example on how the guidedJson or guidedRegex should look like : 
example, if the output of the vLLM has to be : 
```
{
    "name" : "jalotra",
    "country" : "india"
}
```

**Comment by felixzhu555** [comment]
Hi @jalotra this feature is still in development, but once it's added I'd imagine you would define the JSON or regex as a python dictionary or string or pydantic BaseModel class. You can see this simple example that uses pydantic here: [https://github.com/outlines-dev/outlines/blob/main/examples/vllm_integration.py](https://github.com/outlines-dev/outlines/blob/main/examples/vllm_integration.py)

**Comment by felixzhu555** [comment]
#288

**Comment by br3no** [comment]
@felixzhu555 great to see this being picked-up so quickly!

Don't you think the guided decoding in the OpenAI API server should mimic the OpenAI way of having guided decoding? 

Now that outlines is integrated, it becomes easier to support e.g. the `tools` parameter in the chat completion API. 

Take this example request from the OpenAI API reference docs (see https://platform.openai.com/docs/api-reference/chat/create#chat-create-tools): 

```bash
curl https://api.openai.com/v1/chat/completions \
-H "Content-Type: application/json" \
-H "Authorization: Bearer $OPENAI_API_KEY" \
-d '{
  "model": "gpt-3.5-turbo",
  "messages": [
    {
      "role": "user",
      "content": "What is the weather like in Boston?"
    }
  ],
  "tools": [
    {
      "type": "function",
      "function": {
        "name": "get_current_weather",
        "description": "Get the current weather in a given location",
        "parameters": {
          "type": "object",
          "properties": {
            "location": {
              "type": "string",
              "description": "The city and state, e.g. San Francisco, CA"
            },
            "unit": {
              "type": "string",
              "enum": ["celsius", "fahrenheit"]
            }
          },
          "required": ["location"]
        }
      }
    }
  ],
  "tool_choice": "auto"
}'
```

We could extract the json schema from `parameters` and inject a prompt telling the model to generate JSON in this format and at the same time use outlines to enforce the response to match the specification.

**Comment by br3no** [comment]
Supporting `tools` and `tool_choice` would solve #1869.

**Comment by felixzhu555** [comment]
Hey @br3no, thanks for the suggestion, I'll definitely look into supporting the `tools` parameter later on.

**Comment by ibeltagy** [comment]
I gave this a try, and it worked. Thank you

There's a bug in the `dummy_llm`; it failed to find an LLM in the path `dummy`. Here's a simple fix that avoid the need to creating and LLM object because `outlines` doesn't really need it, it only uses the tokenizer
```
    def dummy_llm():
        import types
        x = types.SimpleNamespace()
        y = types.SimpleNamespace()
        x.tokenizer = tokenizer
        y.tokenizer = x
        return y
```

**Comment by ibeltagy** [comment]
@br3no, are you suggesting to remove the proposed `guided_json=schema` option and only support `tools`?

It seems to me that it doesn't have to be one or the other. OpenAI API doesn't allow a custom json schema for constrained decoding, so this is a new capability that's doesn't fit the exiting API. You are right that the guided json can be used to power `tools`, but that doesn't necessarily have to replace `guided_json=schema`.

**Comment by br3no** [comment]
@ibeltagy vLLM offers two server implementations, one of which mimics the OpenAI API. The nice thing about offering an OpenAI-compatible server is that vLLM works as a drop-in replacement for OpenAI. All the stuff that works with OpenAI simply works with vLLM. As an example, take this project here: https://github.com/jxnl/instructor. (I’m not affiliated, nor do I recommend or know much about this project; it’s just an example)

If you start diverging from the OpenAI API, this can break.

At the same time, if users want to use features the OpenAI API doesn’t offer, they are free to use the vLLM-own server implementation, where the vLLM maintainers and community are free to go beyond what OpenAI offers. 

While the OpenAI API doesn’t offer an explicit json schema option, the tools functionality offers a way to do exactly that. 

So yes, I don’t see a good argument for adding these parameters to the OpenAI server because:
1. You can use the json schema functionality through the tools parameters
2. The risk of breaking compatibility gets higher and the cost of maintenance for this server implementation increases 
3. You are free to use the vLLM-own server if you want, having access to json schema and regex-guided decoding.

**Comment by ibeltagy** [comment]
@br3no, makes sense

**Comment by felixzhu555** [comment]
@br3no @ibeltagy the tool parameter is being added through #2488, this PR will only add guided decoding features.

**Review by simon-mo — COMMENTED** [review]
A biiiig problem here is that we are creating new logits processor per request. A common scenarios will be handling for common schemas/constraints. Can you think of a good (and semantically) way to cache the logits processors or the FSM in outlines?

File: tests/entrypoints/test_openai_server_guided_decoding.py
Comment (felixzhu555) [reply tests/entrypoints/test_openai_server_guided_decoding.py:1]: Ok, should the scope for fixtures be reverted to "session"?

**Comment by simon-mo** [comment]
I think module is actually the right one. Otherwise it needs to be in
conftest for uniqueness 

On February 13, 2024, GitHub ***@***.***> wrote:
> @felixzhu555 commented on this pull request.
>

> On tests/entrypoints/test_openai_server_guided_decoding.py
> <https://github.com/vllm-
> project/vllm/pull/2819#discussion_r1488918490>:
>
> Ok, should the scope for fixtures be reverted to "session"?
>

—
Reply to this email directly, view it on GitHub
<https://github.com/vllm-project/vllm/pull/2819#discussion_r1488918490>,
or unsubscribe <https://github.com/notifications/unsubscribe-
auth/AFBD7A2E4CNQHTPJWCDIZRDYTRFNXAVCNFSM6AAAAABDAPCGJ6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQNZZGQYDSNJYHE>.
You are receiving this because you were assigned.Message ID: <vllm-
***@***.***>

File: vllm/model_executor/guided_decoding.py
Comment (lapp0) [reply vllm/model_executor/guided_decoding.py:20]: Maybe we should be improving this on our end and making the logits processor accept a tokenizer as an argument. This would require users to pass `llm.engine.tokenizer.tokenizer` though.

Additionally, `JSONLogitsProcessor` and `RegexLogitsProcessor` aren't currently compatible with vllms `apply_logits_processors` without a patch. However this will be resolved when https://github.com/outlines-dev/outlines/pull/628 is merged.

**Comment by simon-mo** [comment]
Regarding the patch, in this PR we changed our logins processors
handling code gracefully account for the seq_id argument

On February 13, 2024, GitHub ***@***.***> wrote:
> @lapp0 commented on this pull request.
>

> In vllm/model_executor/guided_decoding.py <https://github.com/vllm-
> project/vllm/pull/2819#discussion_r1488951368>:
>
> > +try: + from outlines.serve.vllm import JSONLogitsProcessor,
> RegexLogitsProcessor +except ImportError as e: + raise ValueError( +
> "Please install 'outlines' (pip install outlines) to use guided
> generation." + ) from e + + +class GuidedDecodingMode(Enum): + JSON =
> "json" + REGEX = "regex" + # TODO: add grammar, choice + + +def
> get_guided_decoding_logits_processor(guided_spec: Union[str, dict,
> BaseModel], mode: GuidedDecodingMode, tokenizer): + def dummy_llm():
> Maybe we should be improving this on our end and making the logits
> processor accept a tokenizer as an argument. This would require users
> to pass llm.engine.tokenizer.tokenizer though.
>
> Additionally, JSONLogitsProcessor and RegexLogitsProcessor aren't
> currently compatible with vllms apply_logits_processors without a
> patch. However this will be resolved when outlines-dev/outlines#628
> <https://github.com/outlines-dev/outlines/pull/628> is merged.
>

—
Reply to this email directly, view it on GitHub
<https://github.com/vllm-project/vllm/pull/2819#discussion_r1488951368>,
or unsubscribe <https://github.com/notifications/unsubscribe-
auth/AFBD7AZRLEKEZFUOMTU3Q4DYTRKDDAVCNFSM6AAAAABDAPCGJ6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQNZZGQ2TKOJUGA>.
You are receiving this because you were assigned.Message ID: <vllm-
***@***.***>

File: vllm/model_executor/guided_decoding.py
Comment (br3no) [inline vllm/model_executor/guided_decoding.py:27]: **Suggestion**

I think it would be better to check whether the request contains more than one type of guided decoding parameter and fail instead of defining an implicit ordering.

File: vllm/model_executor/guided_decoding.py
Comment (br3no) [inline vllm/model_executor/guided_decoding.py:83]: **Suggestion**

We could simplify this function by using `re.escape()`.

File: vllm/model_executor/layers/sampler.py
Comment (br3no) [inline vllm/model_executor/layers/sampler.py:157]: **Thought**

We could remove this, since there is no need to pass the `seq_id` anymore, right?

Comment (felixzhu555) [reply vllm/model_executor/layers/sampler.py:157]: I believe we still need to support passing the seq_id in for outlines' logits processors.

File: vllm/model_executor/guided_decoding.py
Comment (felixzhu555) [reply vllm/model_executor/guided_decoding.py:83]: thank you, this is exactly what I was looking for!

Comment (br3no) [reply vllm/model_executor/guided_decoding.py]: **Attention**

`re.escape()` expects a bytes-like object. It will not work with `int`, `float` and `bool`!

File: vllm/model_executor/layers/sampler.py
Comment (br3no) [reply vllm/model_executor/layers/sampler.py:157]: Outlines has accepted a PR that removes this need: https://github.com/outlines-dev/outlines/pull/628

Comment (felixzhu555) [reply vllm/model_executor/layers/sampler.py:157]: I see, I'll revert the change then

File: vllm/model_executor/guided_decoding.py
Comment (br3no) [reply vllm/model_executor/guided_decoding.py:31]: Something like:

```suggestion
    if sum([request.guided_json != None, request.guided_regex != None, request.guided_choice != None]) > 1:
        raise ValueError("You can only use one kind of guided decoding ('json_schema', 'regex' or 'choice').")

    if request.guided_json:
```

File: vllm/model_executor/guided_decoding.py
Comment (simon-mo) [inline vllm/model_executor/guided_decoding.py:75]: * change is_json to enum
* set the cache size to 32 or something, bonus if configurable via env var
* creation of the processor itself should happen in a thread because compilation takes 5-10s for some regex...

Comment (simon-mo) [reply vllm/model_executor/guided_decoding.py:75]: this does mean `get_cached_logits_processor` needs to `async def` which should be okay

File: vllm/model_executor/guided_decoding.py
Comment (br3no) [inline vllm/model_executor/guided_decoding.py:145]: **Suggestion**

Refactor this so that the definition of the `guide` and the `GuidedDecodingMode` are factored out and the `with` scope is reduced to only one call to the executor.

E.g.:

```suggestion
    guide, mode = _get_guide_and_mode(request)

    loop = asyncio.get_running_loop()

    with concurrent.futures.ThreadPoolExecutor(max_workers=2) as pool:
        result = await loop.run_in_executor(
            pool, get_cached_logits_processor,
            guide, tokenizer, mode
        )
        logits_processor = copy(result)
        # reset logits processor's internal state
        logits_processor.fsm_state = defaultdict(int)
        return [logits_processor]
    

def _get_guide_and_mode(request: Union[CompletionRequest, ChatCompletionRequest]) -> tuple[str, GuidedDecodingMode]:
    # validate guided decoding parameters
    guide_count = sum([
        request.guided_json is not None,
        request.guided_regex is not None,
        request.guided_choice is not None
    ])
    if guide_count == 0:
        return None
    elif guide_count > 1:
        raise ValueError(
            "You can only use one kind of guided decoding "
            "('guided_json', 'guided_regex' or 'guided_choice')."
        )
    
    # get the guide and mode
    if request.guided_json:
        if not isinstance(request.guided_json, (str, dict, BaseModel)):
            raise TypeError("JSON schema must be str, dict, or BaseModel")
            
        json = request.guided_json
        if isinstance(json, dict):
            # turn dict into hashable string
            json = json_dumps(json, sort_keys=True)
        elif isinstance(json, BaseModel):
            # use pydantic signature so that different model classes
            # with the same fields will get hashed the same
            json = str(json.__signature__)

        return json, GuidedDecodingMode.JSON
        
    elif request.guided_regex:
        if not isinstance(request.guided_regex, str):
            raise TypeError("Regex must be string")
        
        return request.guided_regex, GuidedDecodingMode.REGEX

    elif request.guided_choice:
        if not isinstance(request.guided_choice, list):
            raise TypeError("Choices must be a list")
        
        # choice just uses regex
        choices = [regex_escape(str(choice)) 
                for choice in request.guided_choice]
        choices_regex = "(" + "|".join(choices) + ")"

        return choices_regex, GuidedDecodingMode.CHOICE
```

**Review by br3no — COMMENTED** [review]
Because outlines logits processors are now compatible with the vLLM logits processors' signature, the changes in `sampler.py` and `sampling_params.py` could all be reverted, I think. This would reduce the changes introduced by this PR.

File: vllm/model_executor/guided_decoding.py
Comment (br3no) [inline vllm/model_executor/guided_decoding.py:26]: **suggestion**

We have added the grammar mode, but there is still no logits processor for grammar guided decoding in outlines (cf. https://github.com/outlines-dev/outlines/blob/main/outlines/serve/vllm.py). While it shouldn't be too difficult to add it there, this will take at least a few days. Maybe we should remove this from the enum and add it in a later PR to avoid bloating this PR? @simon-mo, what do you think?

Comment (felixzhu555) [reply vllm/model_executor/guided_decoding.py:26]: I've added grammar support for now by working off of this [outlines PR](https://github.com/outlines-dev/outlines/pull/541), simon said it should be merged very soon

Comment (br3no) [reply vllm/model_executor/guided_decoding.py:26]: okay!

**Comment by simon-mo** [comment]
We are going to drop grammar support for this PR due to https://github.com/outlines-dev/outlines/issues/685

File: vllm/sampling_params.py
Comment (br3no) [inline vllm/sampling_params.py:1]: The changes here can also all be reverted because we don’t need to support the different signature for the logits processors anymore.

File: vllm/model_executor/guided_decoding.py
Comment (br3no) [inline vllm/model_executor/guided_decoding.py:45]: **nitpick** (no blocker)

Now that the logits processors are part of vLLM, we could introduce a reset_state method to them instead of changing internal fields from outside.

File: vllm/model_executor/layers/sampler.py
Comment (br3no) [inline vllm/model_executor/layers/sampler.py:1]: The same here.

**Comment by br3no** [comment]
We’re very close to the finish line, I’d say.

File: vllm/model_executor/guided_decoding.py
Comment (felixzhu555) [reply vllm/model_executor/guided_decoding.py]: Hey @rlouf, I've copied the logits processor code from outlines over into this PR. Let me know if there's anything you'd like to change to the new file (guided_logits_processors.py).

**Review by br3no — APPROVED** [review]
LGTM!

File: vllm/model_executor/guided_decoding.py
Comment (rlouf) [reply vllm/model_executor/guided_decoding.py]: LGTM

File: tests/entrypoints/test_openai_server.py
Comment (simon-mo) [inline tests/entrypoints/test_openai_server.py:508]: can we check the error code instead?

File: vllm/model_executor/layers/sampler.py
Comment (simon-mo) [reply vllm/model_executor/layers/sampler.py:155]: don't forget to revert the change!

File: vllm/model_executor/guided_logits_processors.py
Comment (simon-mo) [inline vllm/model_executor/guided_logits_processors.py:13]:
```suggestion
# Copyright 2024- the Outlines developers
```

File: vllm/model_executor/guided_logits_processors.py
Comment (simon-mo) [inline vllm/model_executor/guided_logits_processors.py:40]: remove the try except because we are adding it as part of our requirements

File: vllm/model_executor/guided_logits_processors.py
Comment (simon-mo) [inline vllm/model_executor/guided_logits_processors.py:44]: type hint

File: vllm/model_executor/guided_logits_processors.py
Comment (simon-mo) [inline vllm/model_executor/guided_logits_processors.py:43]: can you try adding a unit test for this now we are responsible for maintaining it?

File: vllm/model_executor/guided_decoding.py
Comment (simon-mo) [inline vllm/model_executor/guided_decoding.py:105]: I think this is no longer needed

File: vllm/model_executor/guided_decoding.py
Comment (simon-mo) [inline vllm/model_executor/guided_decoding.py:112]: value error

File: vllm/model_executor/guided_decoding.py
Comment (simon-mo) [inline vllm/model_executor/guided_decoding.py:63]: This validation can be moved to the protocol.py in pydantic model definition. https://docs.pydantic.dev/latest/concepts/validators/#model-validators

File: tests/entrypoints/test_openai_server.py
Comment (felixzhu555) [reply tests/entrypoints/test_openai_server.py:508]: would with `pytest.raises(openai.BadRequestError)` suffice?

Comment (simon-mo) [reply tests/entrypoints/test_openai_server.py:508]: Yup!

File: vllm/model_executor/guided_logits_processors.py
Comment (felixzhu555) [reply vllm/model_executor/guided_logits_processors.py:25]: should it just test the class methods?

Comment (simon-mo) [reply vllm/model_executor/guided_logits_processors.py:25]: Yeah like a new test that create the instance of RegexLogitsProcessor with tokenizer, and verify the input output and fsm transition happens correctly.

File: tests/entrypoints/test_openai_server.py
Comment (rlouf) [inline tests/entrypoints/test_openai_server.py:67]: See second point in [this section of the greenery documentation 
](https://github.com/qntm/greenery?tab=readme-ov-file#unsupported-constructs). `interegular` is a fork of greenery.

Comment (felixzhu555) [reply tests/entrypoints/test_openai_server.py:67]: I see, thanks for clarifying this!

**Comment by creatorrr** [comment]
Slightly late to this thread but imo, [lmfe](https://github.com/noamgat/lm-format-enforcer) is a much better solution to guided decoding than outlines. And it has a vllm integration as well

**Comment by br3no** [comment]
@creatorrr It should be quite easy to support other guided decoding libraries once this PR is merged. I actually had a branch with support for lmfe but could not make the building of the tokenizer prefix tree fast enough. While it only needs to be done once, it takes several tens of seconds. Because of this and because outlines was already in the discussion among the vLLM maintainers, it ended up being the chosen library.

**Comment by creatorrr** [comment]
Gotcha. Makes sense :)

File: tests/entrypoints/test_openai_server.py
Comment (simon-mo) [inline tests/entrypoints/test_openai_server.py:526]: let's make this a new file in tests/entrypoints

File: vllm/model_executor/guided_decoding.py
Comment (simon-mo) [inline vllm/model_executor/guided_decoding.py:37]: This won't work because it won't be a global thread pool. Instead:
* make a global variable called `pool = None`
* if it is `None`, `pool = concurrent.futures.ThreadPoolExecutor(max_workers=2)` ( make sure to declare `global pool`. 
* and then use it

File: vllm/model_executor/guided_decoding.py
Comment (simon-mo) [inline vllm/model_executor/guided_decoding.py:83]: should we just raise ValueError here?

File: vllm/model_executor/guided_logits_processors.py
Comment (simon-mo) [inline vllm/model_executor/guided_logits_processors.py:1]: add the github url for this file for future refernece

**Review by simon-mo — APPROVED** [review]

File: tests/entrypoints/test_openai_server.py
Comment (simon-mo) [inline tests/entrypoints/test_openai_server.py:537]: style issue s/regex/regex_processor

File: tests/entrypoints/test_openai_server.py
Comment (simon-mo) [inline tests/entrypoints/test_openai_server.py:548]: currently this tests the output tensor has the same shape as input tensor
we should also test that output tensor does not equal input tensor. `assert sonLP(token_ids, tensor) == tensor` this will always be true because `tensor` is modified in place. so you do need to make a copy first, also need to use torch.allclose or something to verify `equals`.

File: vllm/model_executor/guided_decoding.py
Comment (felixzhu555) [reply vllm/model_executor/guided_decoding.py:83]: this is the case where the user doesn't pass in any guided decoding params, so I think it shouldn't error

File: vllm/model_executor/guided_decoding.py
Comment (felixzhu555) [inline vllm/model_executor/guided_decoding.py:38]: this pool will never get shutdown while server is up, not sure if that's fine

**Comment by esmeetu** [comment]
@felixzhu555 Great job!
Is it able to support OpenAI `response_format` feature? Link: https://platform.openai.com/docs/api-reference/chat/create
I think it's simple to achieve that by using `guided_json`. cc @simon-mo

**Comment by simon-mo** [comment]
Not quite. The response format requires grammar based sampling because the constraint needs to be arbitrary json. Current the guided json parameter requires a schema.

**Comment by simon-mo** [comment]
Alternatively, if regex works, I don't see why not in a follow up PR. https://stackoverflow.com/questions/2583472/regex-to-validate-json

**Comment by esmeetu** [comment]
> Not quite. The response format requires grammar based sampling because the constraint needs to be arbitrary json. Current the guided json parameter requires a schema. 

Got it! Let's push that feature after this PR merged.

File: vllm/model_executor/guided_decoding.py
Comment (simon-mo) [inline vllm/model_executor/guided_decoding.py:35]:
```suggestion
    if global_pool is None:
```

File: vllm/model_executor/guided_decoding.py
Comment (br3no) [reply vllm/model_executor/guided_decoding.py]: @felixzhu555 have you observed any change in runtime performance after introducing the thread pool? 

I have looked into the outlines/interegular code that implements the FSM building. I’m not sure a separate thread will help here because of the GIL. The code is regular python code and not IO-bound.

@simon-mo this might be precocious optimisation. Or am I missing something here?

File: vllm/model_executor/guided_decoding.py
Comment (simon-mo) [inline vllm/model_executor/guided_decoding.py:21]:
```suggestion
global_thread_pool = None  # used for generating logits processor fsm
```

**Comment by arshadshk** [comment]
Where can I find documentation for using guided decoding?

**Comment by felixzhu555** [comment]
Hi @arshadshk, sorry we don't have written documentation for guided decoding just yet, I'll try to add that soon. If you have a specific use case I can try to explain how to use it, otherwise you can check out the guided decoding tests [here](https://github.com/vllm-project/vllm/blob/b0925b38789bb3b20dcc39e229fcfe12a311e487/tests/entrypoints/test_openai_server.py#L477) for some examples.

**Comment by BedirT** [comment]
Is this strictly an OpenAI-compatible server feature? I don't see any mention on having this available as part of simple `vllm` generation calls? If so what is the reason for it?

**Comment by simon-mo** [comment]
Currently yes. But it would be valuable to have similar API is `LLM` class. I will open a new issue to have folks working on it.

**Comment by simon-mo** [comment]
@BedirT issue created #3536. I will have someone working on it.

**Comment by ProVega** [comment]
Where can I find documentation for this feature?

**Comment by BedirT** [comment]
I don't think there is a lengthy documentation on it, but it is shortly mentioned here: https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html#extra-parameters-for-completions-api

**Comment by ProVega** [comment]
That same states the parameters, but doesn't provide any actual samples of JSON SCHEMA. I have found a few others, but I can't find anything that shows how to return an ARRAY of JSON objects. Example: "Find me 5 companies that sell cars" and I get back { Results: [ { name: "Ford" }, {name: "Toyota"}, {name: "BMW}]

**Comment by AaronFriel** [comment]
@ProVega Using tools like pydantic or zod-to-json-schema are helpful here. I think the schema you want is:

```json
{
  "type": "array",
  "items": {
    "type": "object",
    "properties": {
      "name": { "type": "string" }
    }
  }
}
```

**Comment by sam1am** [comment]
This took me some tinkering to figure out, given all the different methods and tools that have been discussed regarding this issue. Here is a simple example that will produce guided json output for me consistently: 

```
from openai import OpenAI
import json
import time

client = OpenAI(
    base_url="http://localhost:8000/v1",
)

json_schema = {
    "type": "object",
    "properties": {
        "thought": {"type": "string"},
        "answer": {"type": "string"}
    },
    "required": ["thought", "answer"]
}

query = "What is the capital of France?"

# Feed the random question into the existing query
system_prompt = "Respond only with a json object containing the following fields and nothing else: thought, answer."
completion = client.chat.completions.create(
    model="casperhansen/llama-3-8b-instruct-awq",
    messages=[
        {"role": "system", "content": system_prompt},
        {"role": "user", "content": query}
    ],
    extra_body={
        "stop_token_ids": [128009], 
        "response_format": {"type": "json_object"},
        "guided_json": json_schema
    }
)

print(completion.choices[0].message.content.strip())

```
Output:
```
{
"answer": "Paris",
"thought": "Bonjour!" 
}
```

The only issue is the order, which appears to be alphabetical regardless of how I write the code. This is important for tree of thought-type prompting.

**Pull request merged**
